lib/spack/spack/package.py

```diff
@@ -1505,9 +1521,9 @@ def do_install(self, **kwargs):
             # module from the upstream Spack instance.
             return
 
+        # Ensure package code is not already installed
```

Comment:
This check is actually to see if there was some leftover code from an unsuccessful installation. I think it would be more accurate to move this to the `elif layout.check_installed(self.spec)` check below. Or remove it, because IMO that is fairly self-explanatory from the code which surrounds that check.

Reply:
I'm not sure I follow your issue with the existing code.  Perhaps the method name is misleading since it has side-effects.

I could see checking the result of the check before taking a read lock and checking the layout.

Reply:
Originally the comment appeared right before the `if` statement which included the relevant code:

```
        # Ensure package is not already installed
        layout = spack.store.layout
        with spack.store.db.prefix_read_lock(self.spec):
            ...
            if layout.check_installed(self.spec):
```

Now things are relocated and it appears right before an unrelated check:

```
        # Ensure package is not already installed
        partial = self.check_for_unfinished_installation(keep_prefix, restage)
```

I think the comment is mainly relevant to the `check_installed` call.

Reply:
Ah.  Thanks for clarifying.  I misunderstood the issue and now see I was "conflating" the checks by moving the comment.   Let me look at that again and make sure the comments are appropriate.